### PR TITLE
Centralizes images into agnhost (part 1)

### PR DESCRIPTION
**Pull request by claudiubelu**
**What type of PR is this?**

/kind feature

/sig testing

**What this PR does / why we need it**:

Refactors the functions used in agnhost into different modules, based on their functionality, leaving only the main in the base folder.

Future commits will add several functionalities to agnhost, so this change will be necessary to keep it clean.

Centralizes the following images into agnhost:

- fakegitserver
- hostexec
- liveness
- logs-generator
- no-snat-test
- no-snat-test-proxy
- port-forward-tester

**Which issue(s) this PR fixes**:

Related: #76342

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

**Comment by listx** [comment]
Looks like you are consolidating the separate Go modules into 1, just as this PR says. Sounds good to me.

/lgtm

@mkumatag Do you have any additional thoughts on this?

@bclau Should the people in #76342 also take a few minutes to review this PR? I ask because I only have a high-level understanding of what you are doing here and can't speak for the code changes.

**Comment by bclau** [comment]
> 
> 
> Looks like you are consolidating the separate Go modules into 1, just as this PR says. Sounds good to me.
> 
> /lgtm
> 
> @mkumatag Do you have any additional thoughts on this?
> 
> @bclau Should the people in #76342 also take a few minutes to review this PR? I ask because I only have a high-level understanding of what you are doing here and can't speak for the code changes.

Any feedback is welcome, of course. As for the changes I've introduced in this PR, those are the most significant:

- I've mostly used ``git mv``, so the actual changes are easier to observe.
- I've removed ``go_binary`` from the ``BUILD`` files, since we're centralizing everything into the ``agnhost`` binary. We could leave it in, and make each be compilable on its own, if needed, but I don't know if we want that anymore.
- "refactored" the various flags into cobra flags, so they show up under their subcommand's help (e.g.: ``./agnhost help logs-generator``)
- moved any ``README``s to the ``agnhost``'s ``README`` and created new sections for the ones that didn't have any. I did my best to make them as helpful as possible, but if I've missed any important information, please let me know.

**Comment by listx** [comment]
> > Looks like you are consolidating the separate Go modules into 1, just as this PR says. Sounds good to me.
> > /lgtm
> > @mkumatag Do you have any additional thoughts on this?
> > @bclau Should the people in #76342 also take a few minutes to review this PR? I ask because I only have a high-level understanding of what you are doing here and can't speak for the code changes.
> 
> Any feedback is welcome, of course. As for the changes I've introduced in this PR, those are the most significant:
> 
> * I've mostly used `git mv`, so the actual changes are easier to observe.
> * I've removed `go_binary` from the `BUILD` files, since we're centralizing everything into the `agnhost` binary. We could leave it in, and make each be compilable on its own, if needed, but I don't know if we want that anymore.
> * "refactored" the various flags into cobra flags, so they show up under their subcommand's help (e.g.: `./agnhost help logs-generator`)
> * moved any `README`s to the `agnhost`'s `README` and created new sections for the ones that didn't have any. I did my best to make them as helpful as possible, but if I've missed any important information, please let me know.
> 
> As far as testing goes, I've already built `claudiubelu/agnhost:1.1` with all the images enumerated under part 1 and part 2 (and a few others) centralized into it, and replaced those images with `agnhost` in tests, and as far as Conformance tests go, they've passed for me.

These changes sound fine to me, but I'd like to ask for others' opinions as well if possible before I approve as an owner.

Also, given the amount of activity you have in this area, it might be worth trying to add you as an approver (even though you would still need others' approvals in PRs). Are you a kubernetes org member yet?

**Comment by bclau** [comment]
> 
> These changes sound fine to me, but I'd like to ask for others' opinions as well if possible before I approve as an owner.
> 
> Also, given the amount of activity you have in this area, it might be worth trying to add you as an approver (even though you would still need others' approvals in PRs). Are you a kubernetes org member yet?

I am: https://github.com/kubernetes/org/issues/509

**Comment by listx** [comment]
@mkumatag @luxas Thoughts?

File: test/images/agnhost/VERSION
Comment (mkumatag) [inline test/images/agnhost/VERSION]: Let us bump up to the major release because a lot of changes went in(something like 2.0)

**Review by mkumatag — CHANGES_REQUESTED** [review]

File: test/images/agnhost/README.md
Comment (mkumatag) [inline test/images/agnhost/README.md]: I'm not really sure what this `agnhost` means but question is - is this name still relevant after adding so many other subcommands?

**Review by mkumatag — CHANGES_REQUESTED** [review]

**Comment by mkumatag** [comment]
> @mkumatag @luxas Thoughts?

overall changes looks okay to me, lets do the following thing to make sure new changes works fine!

@listx push the new image 
@bclau let us submit the commit with using above image in manifest.go and tests once @listx pushes the image to make sure all the conformance passes.

**Comment by bclau** [comment]
> 
> 
> > @mkumatag @luxas Thoughts?
> 
> overall changes looks okay to me, lets do the following thing to make sure new changes works fine!
> 
> @listx push the new image
> @bclau let us submit the commit with using above image in manifest.go and tests once @listx pushes the image to make sure all the conformance passes.

Already sent the PR for that: https://github.com/kubernetes/kubernetes/pull/78394

**Comment by mkumatag** [comment]
> > > @mkumatag @luxas Thoughts?
> > 
> > 
> > overall changes looks okay to me, lets do the following thing to make sure new changes works fine!
> > @listx push the new image
> > @bclau let us submit the commit with using above image in manifest.go and tests once @listx pushes the image to make sure all the conformance passes.
> 
> Already sent the PR for that: #78394

Thanks @bclau, @listx can you build and push new `agnhost` image to pass other PR?

File: test/images/agnhost/README.md
Comment (bclau) [reply test/images/agnhost/README.md]: Eh, less so. So, the name ``agnhost`` comes from "agnhost host", which basically means that it'll work the same on any OS. But it is still true, the image still contains the common Linux utilities on both Linux and Windows, including /bin/sh, nc, etc., and 3 of the current subcommands still accomplish the original goal. Adding the other subcommands doesn't necessarely go against its goal.

File: test/images/agnhost/VERSION
Comment (bclau) [reply test/images/agnhost/VERSION]: Done

**Comment by listx** [comment]
@mkumatag I'll try to get the images pushed later today, caught up with something at the moment.

Sorry about the delay!

**Comment by listx** [comment]
I pushed the images. Log below:

```
./image-util.sh push agnhost
The push refers to repository [gcr.io/kubernetes-e2e-test-images/agnhost-amd64]
5e365e26c3ca: Pushed
3aa628b98439: Pushed
721384ec99e5: Layer already exists
2.0: digest: sha256:2c0b5d8e1da6bb525f0d4b51c220dac7e8d4081edab7ef728e7931827390eabc size: 950                                                                                    
The push refers to repository [gcr.io/kubernetes-e2e-test-images/agnhost-arm]
a88e19706e8a: Pushed
bfa597a7eed6: Pushed
350c5d4fce68: Pushed
845b5c25f81e: Layer already exists
2.0: digest: sha256:fdd877012a1c31ff6cc6728cf30911bb77cb7119a127af86249deaaf63a6aadc size: 1161                                                                                   
The push refers to repository [gcr.io/kubernetes-e2e-test-images/agnhost-arm64]
5cc225da45e7: Pushed
15901ef31942: Pushed
874c57655f97: Pushed
56e86ee0b452: Layer already exists
2.0: digest: sha256:6ac5972b281c12dbb9aca99b8e498de443539be0ccc7fb0f470cd166ff936d28 size: 1161                                                                                   
The push refers to repository [gcr.io/kubernetes-e2e-test-images/agnhost-ppc64le]
2396d6e23399: Pushed
109e2b3d6de0: Pushed
796f1353ab3d: Pushed
85e899b8f0cc: Layer already exists
2.0: digest: sha256:1c58fa90acc5fd9e5577cde4564de3c1d2426daf6cc3c0d95cd2a3182fd547e8 size: 1161                                                                                   
The push refers to repository [gcr.io/kubernetes-e2e-test-images/agnhost-s390x]
021229db5da7: Pushed
16bc4535b331: Pushed
f129ff7d977b: Layer already exists
2.0: digest: sha256:ac24934dcaf3e0d274fefe18abc69c5f91d074b2fd672f44f18523e0d130ac26 size: 950                                                                                    
Created manifest list gcr.io/kubernetes-e2e-test-images/agnhost:2.0
Pushed ref gcr.io/kubernetes-e2e-test-images/agnhost@sha256:2c0b5d8e1da6bb525f0d4b51c220dac7e8d4081edab7ef728e7931827390eabc with digest: sha256:2c0b5d8e1da6bb525f0d4b51c220dac7e8d4081edab7ef728e7931827390eabc
Pushed ref gcr.io/kubernetes-e2e-test-images/agnhost@sha256:fdd877012a1c31ff6cc6728cf30911bb77cb7119a127af86249deaaf63a6aadc with digest: sha256:fdd877012a1c31ff6cc6728cf30911bb77cb7119a127af86249deaaf63a6aadc
Pushed ref gcr.io/kubernetes-e2e-test-images/agnhost@sha256:6ac5972b281c12dbb9aca99b8e498de443539be0ccc7fb0f470cd166ff936d28 with digest: sha256:6ac5972b281c12dbb9aca99b8e498de443539be0ccc7fb0f470cd166ff936d28
Pushed ref gcr.io/kubernetes-e2e-test-images/agnhost@sha256:1c58fa90acc5fd9e5577cde4564de3c1d2426daf6cc3c0d95cd2a3182fd547e8 with digest: sha256:1c58fa90acc5fd9e5577cde4564de3c1d2426daf6cc3c0d95cd2a3182fd547e8
Pushed ref gcr.io/kubernetes-e2e-test-images/agnhost@sha256:ac24934dcaf3e0d274fefe18abc69c5f91d074b2fd672f44f18523e0d130ac26 with digest: sha256:ac24934dcaf3e0d274fefe18abc69c5f91d074b2fd672f44f18523e0d130ac26
sha256:75990e17b63596dc69c3fd3161a281445048457da1555caebfecd5c1215a7c2e
```

**Comment by bclau** [comment]
> 
> 
> I pushed the images. Log below:
> 

Thanks! It seems that the tests have passed: https://github.com/kubernetes/kubernetes/pull/78389

**Comment by timothysc** [comment]
/assign @dims 

Who on the google side do we need to loop in here?

**Comment by mkumatag** [comment]
> /assign @dims
> 
> Who on the google side do we need to loop in here?

@listx already working on this.

**Comment by mkumatag** [comment]
I see all green here - https://github.com/kubernetes/kubernetes/pull/78394, I think we are good.

/lgtm

File: test/images/agnhost/README.md
Comment (spiffxp) [inline test/images/agnhost/README.md]:
```suggestion
The image can be found at `gcr.io/kubernetes-e2e-test-images/agnhost:2.0` for Linux
```

File: test/images/agnhost/README.md
Comment (spiffxp) [inline test/images/agnhost/README.md]:
```suggestion
containers, and `e2eteam/agnhost:2.0` for Windows containers. In the future, the same
```

File: test/images/agnhost/Dockerfile
Comment (spiffxp) [inline test/images/agnhost/Dockerfile]: non-blocking nit: I can't tell that these are actually consumed by `log_generator.go` so it's unclear why they're necessary

File: test/images/agnhost/README.md
Comment (spiffxp) [inline test/images/agnhost/README.md]: this paragraph is incorrect: near as I can tell, `logs_generator.go` only takes flags, not env vars, and I can't tell that agnhost plumbs the env vars through?

File: test/images/logs-generator/Dockerfile
Comment (spiffxp) [inline test/images/logs-generator/Dockerfile]: this explains my confusion: the old logs-generator container uses env vars to pass to the parameters; the agnhost refactor does not

**Review by spiffxp — CHANGES_REQUESTED** [review]
/lgtm cancel
this is really close I just have some tiny nits

File: test/images/logs-generator/Dockerfile
Comment (johnSchnake) [reply test/images/logs-generator/Dockerfile]: Yea the new one just uses flags. Based on https://github.com/kubernetes/kubernetes/pull/78394/files#diff-7dc1ebd3b2f2f6ca1ff248e6601016a6R108 it looks like the invocation was updated to use the flags instead of env vars too, so the doc notes about the env vars can be removed.

File: test/images/agnhost/Dockerfile
Comment (bclau) [reply test/images/agnhost/Dockerfile]: True, it seems that the env vars are not consumed, the logs-generator binary was expecting those params as flags from the initial commit, yet the documentation was also talking about env vars. The tests also sets those env vars.

File: test/images/agnhost/README.md
Comment (bclau) [reply test/images/agnhost/README.md]: This part of the documentation was pretty much copied from the original documentation logs-generator had. Interestingly, it was like this from the beggining: accepting those parameters as flags and not as env vars, yet the documentation is talking about env vars. The tests also sets those env vars. In the replacement PR, I've replaced that with actual flags: https://github.com/kubernetes/kubernetes/pull/78394/files#diff-7dc1ebd3b2f2f6ca1ff248e6601016a6R108

File: test/images/agnhost/README.md
Comment (bclau) [reply test/images/agnhost/README.md]: Oups, I forgot to update the version in the documentation.

File: test/images/agnhost/README.md
Comment (bclau) [reply test/images/agnhost/README.md]: ditto.

File: test/images/logs-generator/Dockerfile
Comment (bclau) [reply test/images/logs-generator/Dockerfile]: Good catch, there's no longer a need for those env variables. I'll also remove them from the Dockerfile.

**Comment by bclau** [comment]
I've updated the README with the new usage and removed the env vars from the Dockerfile.

**Review by spiffxp — APPROVED** [review]
/lgtm
thanks for the updates

File: test/images/agnhost/logs-generator/logs_generator.go
Comment (johnSchnake) [inline test/images/agnhost/logs-generator/logs_generator.go]: Shouldnt this be 0 args; the variables are given via flags not arguments.

Comment (bclau) [reply test/images/agnhost/logs-generator/logs_generator.go]: Oups, you're right. Then that means that pretty much all commands should have this at 0.

**Pull request merged**
